galaxy/tools/verify/interactor.py

```diff
@@ -635,44 +675,58 @@ def _verify_composite_datatype_file_content(file_name, hda_id, base_name=None, a
             keep_outputs_dir=keep_outputs_dir,
+            mode=mode,
         )
     except AssertionError as err:
         errmsg = 'Composite file (%s) of %s different than expected, difference:\n' % (base_name, item_label)
         errmsg += str(err)
         raise AssertionError(errmsg)
 
 
-def _verify_extra_files_content(extra_files, hda_id, dataset_fetcher, test_data_path_builder, keep_outputs_dir):
+def _verify_extra_files_content(extra_files, hda_id, dataset_fetcher, test_data_downloader, keep_outputs_dir):
     files_list = []
+    cleanup_directories = []
     for extra_file_dict in extra_files:
         extra_file_type = extra_file_dict["type"]
         extra_file_name = extra_file_dict["name"]
         extra_file_attributes = extra_file_dict["attributes"]
         extra_file_value = extra_file_dict["value"]
 
         if extra_file_type == 'file':
-            files_list.append((extra_file_name, extra_file_value, extra_file_attributes))
+            files_list.append((extra_file_name, extra_file_value, extra_file_attributes, extra_file_type))
         elif extra_file_type == 'directory':
-            for filename in os.listdir(test_data_path_builder(extra_file_value)):
-                files_list.append((filename, os.path.join(extra_file_value, filename), extra_file_attributes))
+            extracted_path = test_data_downloader(extra_file_value, mode='directory')
+            cleanup_directories.append(extracted_path)
+            for root, directories, files in util.path.safe_walk(extracted_path):
+                for filename in files:
+                    filename = os.path.join(root, filename)
+                    filename = os.path.relpath(filename, extracted_path)
+                    files_list.append((filename, os.path.join(extracted_path, filename), extra_file_attributes, extra_file_type))
         else:
             raise ValueError('unknown extra_files type: %s' % extra_file_type)
-    for filename, filepath, attributes in files_list:
-        _verify_composite_datatype_file_content(filepath, hda_id, base_name=filename, attributes=attributes, dataset_fetcher=dataset_fetcher, test_data_path_builder=test_data_path_builder, keep_outputs_dir=keep_outputs_dir)
+    try:
+        for filename, filepath, attributes, extra_file_type in files_list:
+            _verify_composite_datatype_file_content(filepath, hda_id, base_name=filename, attributes=attributes, dataset_fetcher=dataset_fetcher, test_data_downloader=test_data_downloader, keep_outputs_dir=keep_outputs_dir, mode=extra_file_type)
+    finally:
+        for path in cleanup_directories:
+            shutil.rmtree(path)
 
 
-def verify_tool(tool_id, galaxy_interactor, resource_parameters={}, register_job_data=None, test_index=0, tool_version=None, quiet=False, force_path_paste=False):
+def verify_tool(tool_id, galaxy_interactor, resource_parameters=None, register_job_data=None, test_index=0, tool_version=None, quiet=False, test_history=None):
```

Comment:
`force_path_paste=False` should be re-added, before or after `test_history=None`